LEGO1/mxhashtable.h

```diff
@@ -93,77 +75,92 @@ class MxHashTable : protected HashTableParent<T> {
 template <class T>
 class MxHashTableCursor : public MxCore {
 public:
-	MxHashTableCursor(MxHashTable<T>* p_hashTable)
+	MxHashTableCursor(MxHashTable<T>* p_table)
 	{
-		m_table = p_hashTable;
+		m_table = p_table;
 		m_match = NULL;
 	}
 
-	MxBool Find(T* p_obj)
-	{
-		MxU32 hash = m_table->Hash(p_obj);
-		int bucket = hash % m_table->m_numSlots;
+	MxBool Find(T p_obj);
+	MxBool Current(T& p_obj);
+	void DeleteMatch();
 
-		MxHashTableNode<T>* t = m_table->m_slots[bucket];
+private:
+	MxHashTable<T>* m_table;
+	MxHashTableNode<T>* m_match;
+};
 
-		while (t) {
-			if (t->m_hash == hash && !m_table->Compare(t->m_obj, p_obj))
-				m_match = t;
-			t = t->m_next;
-		}
+template <class T>
+MxBool MxHashTableCursor<T>::Find(T p_obj)
+{
+	MxU32 hash = m_table->Hash(p_obj);
+	MxS32 bucket = hash % m_table->m_numSlots;
 
-		return m_match != NULL;
-	}
+	MxHashTableNode<T>* t = m_table->m_slots[bucket];
 
-	void GetMatch(T*& p_obj)
-	{
-		if (m_match) {
-			p_obj = m_match->m_obj;
-		}
+	while (t) {
+		if (t->m_hash == hash && !m_table->Compare(t->m_obj, p_obj))
+			m_match = t;
+		t = t->m_next;
 	}
 
-	void DeleteMatch()
-	{
-		// Cut the matching node out of the linked list
-		// by updating pointer references.
+	return m_match != NULL;
+}
 
-		if (m_match->m_prev) {
-			m_match->m_prev->m_next = m_match->m_next;
-		}
-		else {
-			// No "prev" node, so move "next" to the head of the list.
-			int bucket = m_match->m_hash % m_table->m_numSlots;
-			m_table->m_slots[bucket] = m_match->m_next;
-		}
+template <class T>
+MxBool MxHashTableCursor<T>::Current(T& p_obj)
+{
+	if (m_match) {
+		p_obj = m_match->m_obj;
+	}
 
-		if (m_match->m_next)
-			m_match->m_next->m_prev = m_match->m_prev;
+	return m_match != NULL;
+}
 
-		m_table->m_customDestructor(m_match->m_obj);
-		delete m_match;
-		m_table->m_numKeys--;
+template <class T>
+void MxHashTableCursor<T>::DeleteMatch()
```

Comment:
I think the whole function body is wrapped in a `if (m_match) {}` test in the alpha build (maybe you can double check). I think this check is optimized away by the compiler because it can deduce that `m_match` is not `NULL` in the one case the function is used in our decomp, so even if we add it the assembly doesn't change. iirc the match percentages of some other functions changed though. Personally I'm on the fence if we should just add it as well to match the alpha debug code, since it's obviously not guaranteed to be the same as the original final source, but I guess we could add a code comment about it at least.

Reply:
Yep, it's there (at `LEGO1D 0x1008d8bf`). I'll put that in. Match percentage seems ok, except for `MxHashTable<MxVariable *>::Resize` but that must be compiler weirdness.